subcommands/targets/add.go

```diff
@@ -0,0 +1,288 @@
+)
+
+func NewTarget(tufMeta tuf.FileMeta) (*Target, error) {
+	custom, err := api.TargetCustom(tufMeta)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal Target custom from TUF meta: %s", err.Error())
+	}
+	return &Target{
+		Length: tufMeta.Length,
+		Hashes: tufMeta.Hashes,
+		Custom: custom,
+	}, nil
+}
+
+func (t *Target) Derive(newVer int) *Target {
+	derivedCustom := *t.Custom
+	derivedCustom.Version = strconv.Itoa(newVer)
+	derivedCustom.CreatedAt = time.Now().UTC().Format(time.RFC3339)
+	derivedCustom.UpdatedAt = derivedCustom.CreatedAt
+	return &Target{
+		Length: t.Length,
+		Hashes: t.Hashes,
+		Custom: &derivedCustom,
+	}
+}
+
+func (t *Target) HardwareId() string {
+	return t.Custom.HardwareIds[0]
+}
+
+func (t *Target) Version() int {
+	ver, err := strconv.Atoi(t.Custom.Version)
+	if err != nil {
+		panic(fmt.Sprintf("failed to convert Target version to int: %s", err.Error()))
+	}
+	return ver
+}
+
+func (t *Target) Tags() []string {
+	return t.Custom.Tags
+}
+
+func (t *Target) HasTag(tag string) bool {
+	for _, ct := range t.Tags() {
+		if ct == tag {
+			return true
+		}
+	}
+	return false
+}
+
+func (t *Target) Name() string {
+	return t.Custom.Name + "-" + t.Custom.Version
+}
+
+func (t *Target) SetHash(hash string) error {
+	hb, err := base64.StdEncoding.DecodeString(hash)
+	if err != nil {
+		return err
+	}
+	t.Hashes[notary.SHA256] = hb
+	return nil
+}
+
+func init() {
+	addCmd := &cobra.Command{
+		Use:   "add",
+		Short: "Add target(s)",
+		Run:   doAdd,
+		Long:  "",
+	}
+	cmd.AddCommand(addCmd)
+	addCmd.Flags().StringVarP(&addTargetType, "type", "", "", "Target type")
+	addCmd.Flags().StringVarP(&addTags, "tags", "", "", "comma,separate,list of Target tags")
+	addCmd.Flags().StringVarP(&addSrcTag, "src-tag", "", "", "OSTree Target tag to base app targets on")
+	addCmd.Flags().StringVarP(&addOutFile, "out", "o", "", "a file path to output the new Targets to")
+	addCmd.Flags().BoolVarP(&addPrint, "print", "", false, "print generated new Targets to stdout")
+	addCmd.Flags().BoolVarP(&addDryRun, "dry-run", "", false, "don't post generated new Targets")
+}
+
+func doAdd(cmd *cobra.Command, args []string) {
+	factory := viper.GetString("factory")
+	supportedTargetTypes := map[string]func(factory string, tags []string, srcTag string, args []string) (Targets, error){
+		"app":    createAppTargets,
+		"ostree": createOstreeTarget,
+	}
+	if len(addTargetType) == 0 {
+		subcommands.DieNotNil(errors.New("missing mandatory flag `--type`"))
+	}
+	deriveTargets, ok := supportedTargetTypes[addTargetType]
+	if !ok {
+		subcommands.DieNotNil(errors.New("unsupported type of Target: " + addTargetType))
+	}
+	if len(addTags) == 0 {
+		subcommands.DieNotNil(errors.New("missing mandatory flag `--tags`"))
+	}
+	tags := strings.Split(addTags, ",")
+	if len(addSrcTag) == 0 {
+		subcommands.DieNotNil(errors.New("missing mandatory flag `--src-tag`"))
+	}
+
+	newTargets, err := deriveTargets(factory, tags, addSrcTag, args)
```

Comment:
I thought about it. There is a quite overhead with adding new sub-command, also this two commands share 90-95% of the code. So, I am not sure that it's justifiable atm. Maybe if we get users for it and get feedback and/or each of the commands will need more specific flags then it can be justified, IMO.

Reply:
you can share the code and still have separate subcommands. However, if you like the current ergonomics of the command, we can keep it.

Reply:
I still need that `if !ok` since the `supportedTargetTypes` map can be incorrectly defined.